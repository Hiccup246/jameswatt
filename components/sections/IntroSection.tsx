import { useEffect, useRef, useState } from "react";
import { WaveTank } from "../WaveTank";
import Image from "next/image";
import CircleDripIcon from "../Icons/CircleDripIcon";
import JamesWattImage from "../../public/panthy-tiny.webp";
import Human from "../../public/human.webp";
import Developer from "../../public/developer.webp";

// Source code for the functions easeInCirc, updateJuice, update, resize, drop, useEffect and the
// related SVG code is courtesy of the Fresh framework website. The code source can
// be found here https://github.com/denoland/fresh/tree/main/www

function easeInCirc(x: number) {
  return 1 - Math.sqrt(1 - Math.pow(x, 2));
}

const waveTank = new WaveTank();

const SVG_WIDTH = 100;
const DROP_OFFSET = 82;

export default function IntroSection() {
  const [counter, setCounter] = useState(0);
  const [dropy, setDropy] = useState(10);
  const [width, setWidth] = useState(SVG_WIDTH);
  const [springs, setSprings] = useState<Spring[]>(waveTank.springs);

  const widthRef = useRef(width);
  const requestIdRef = useRef<number>();

  const grid = SVG_WIDTH / waveTank.waveLength;
  const points = [
    [0, 100],
    [0, 0],
    ...springs.map((x, i) => [i * grid, x.position]),
    [width, 0],
    [width, 100],
  ];
  const springsPath = `${points.map((x) => x.join(",")).join(" ")}`;

  function updateJuice(timestamp: number) {
    const amp = 40;
    const x = timestamp / 2000;
    const saw = x - Math.floor(x);
    if (saw < 0.6) {
      setCounter(easeInCirc(saw) * amp);
      setDropy(10);
    } else {
      setCounter(easeInCirc(1 - saw) * amp * 0.1);
      setDropy(10 + Math.pow(saw - 0.6, 2) * 10000);
    }
  }

  function update(timestamp: number) {
    updateJuice(timestamp);
    waveTank.update(waveTank.springs);
    setSprings([...waveTank.springs]);

    const offset = 500;
    const saw =
      (timestamp + offset) / 2000 - Math.floor((timestamp + offset) / 2000);
    if (saw < 0.01) {
      drop();
    }
    requestIdRef.current = globalThis.requestAnimationFrame(update);
  }

  function resize() {
    setWidth(document.body.clientWidth);
  }

  function drop() {
    const dropPosition = Math.round(
      ((widthRef.current / 2 + DROP_OFFSET) / widthRef.current) * 100
    );
    waveTank.springs[dropPosition].position = -60;
  }

  useEffect(() => {
    widthRef.current = width;
  }, [width]);

  useEffect(() => {
    const mediaQuery = window.matchMedia("(prefers-reduced-motion: reduce)");
    if (mediaQuery.matches) {
      return;
    }

    startAnimation();
    globalThis.addEventListener("resize", resize);
    resize();

    return () => {
      globalThis.removeEventListener("resize", resize);
      if (requestIdRef.current !== undefined) {
        cancelAnimationFrame(requestIdRef.current);
      }
    };
  }, []);

  function startAnimation() {
    requestIdRef.current = requestAnimationFrame(update);
  }

  return (
    <section className="flex h-full w-full flex-col bg-brown dark:bg-darkgrey">
      <div className="grid h-fit w-full grid-cols-1 md:mb-10 md:grid-cols-2 lg:mt-12">
        <div className="relative flex h-auto w-full flex-col justify-start md:justify-start md:pr-10 lg:pr-28">
          <Image
            src={Human}
            priority
            height={80}
            className="place-self-center object-scale-down md:place-self-end"
            alt="Human Text"
          />
        </div>

        <div
          className={`
               relative flex h-auto w-full flex-col
               justify-start md:justify-start md:pl-10 lg:pl-28
        `}
        >
          <Image
            src={Developer}
            priority
            height={80}
            className="place-self-center object-scale-down md:place-self-start"
            alt="Developer Text"
          />
        </div>
      </div>

      <div className="relative mx-auto mb-20 h-[372.5px] w-[310px] bg-brown md:mb-28 lg:mb-16 xl:mb-36">
        <div className="absolute -top-px -left-px z-20 h-[373.5px] h-full w-[311px] w-full stroke-white text-brown dark:stroke-redgrey dark:text-darkgrey">
          <CircleDripIcon />
        </div>

        <div className="absolute top-0 left-0 z-10 h-full w-full text-brown">
          <Image
            src={JamesWattImage}
            height={372.5}
            width={310}
            priority
            placeholder="blur"
            alt="James Watt"
            sizes="310px"
          />
        </div>

        <svg
          width="100"
          height="350"
          viewBox="0 0 100 200"
          className="absolute left-[216.5px] top-[184px] z-20"
          aria-label="Water Drop"
        >
          <circle
            cx="24"
            cy={dropy}
            r="6"
            className="fill-white dark:fill-redgrey"
          ></circle>
        </svg>

        <svg
          className="absolute left-[232px] top-[215px] z-20"
          width="16.5"
          aria-label="Growing Drop"
        >
          <path
            d={`M0,48 a1,${1.3 + (counter + 1) * 0.2} 0 0,0 16.5,0`}
            className="fill-white dark:fill-redgrey"
          />
        </svg>
      </div>

      <svg
        width="100%"
        height="100px"
        viewBox="0 0 100 100"
        preserveAspectRatio="none"
        className="-mt-5"
        aria-label="Wave Tank"
      >
        <polygon
          points={springsPath}
          className="fill-white dark:fill-redgrey dark:stroke-redgrey"
          stroke="white"
          transform="translate(0, 50)"
        ></polygon>
      </svg>
    </section>
  );
}
