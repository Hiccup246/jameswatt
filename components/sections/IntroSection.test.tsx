import { render } from "@testing-library/react";
import IntroSection from "./IntroSection";

describe("IntroSection", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      configurable: true,
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
      })),
    });
  });

  it("renders correctly", () => {
    const { container } = render(<IntroSection />);

    expect(container).toMatchSnapshot();
  });
});
