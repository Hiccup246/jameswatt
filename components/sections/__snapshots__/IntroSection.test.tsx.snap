// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntroSection renders correctly 1`] = `
<section
  className="flex h-full w-full flex-col bg-brown dark:bg-darkgrey"
>
  <div
    className="
             grid h-fit w-full grid-cols-1
             md:mb-10 md:grid-cols-2 lg:mt-12
      "
  >
    <div
      className="
               relative flex h-auto w-full flex-col
               justify-start md:justify-start md:pr-10 lg:pr-28
        "
    >
      <img
        alt="Human Text"
        className="place-self-center object-scale-down md:place-self-end"
        data-nimg="1"
        decoding="async"
        height={80}
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fimg.jpg&w=256&q=75"
        srcSet="/_next/image?url=%2Fimg.jpg&w=96&q=75 1x, /_next/image?url=%2Fimg.jpg&w=256&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={80}
      />
    </div>
    <div
      className="
               relative flex h-auto w-full flex-col
               justify-start md:justify-start md:pl-10 lg:pl-28
        "
    >
      <img
        alt="Developer Text"
        className="place-self-center object-scale-down md:place-self-start"
        data-nimg="1"
        decoding="async"
        height={80}
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fimg.jpg&w=256&q=75"
        srcSet="/_next/image?url=%2Fimg.jpg&w=96&q=75 1x, /_next/image?url=%2Fimg.jpg&w=256&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={80}
      />
    </div>
  </div>
  <div
    className="relative mx-auto mb-20 h-[376px] w-[310px] bg-brown md:mb-28 lg:mb-16 xl:mb-36"
  >
    <div
      className="absolute top-0 left-0 z-20 h-full w-full stroke-white text-brown dark:stroke-redgrey dark:text-darkgrey"
    >
      <svg
        className=""
        viewBox="0 0 569 695"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M 0.00,0.00 C 0.00,0.00 569.00,0.00 569.00,0.00 569.00,0.00 569.00,695.00 569.00,695.00 569.00,695.00 0.00,695.00 0.00,695.00 0.00,695.00 0.00,0.00 0.00,0.00 Z M 278.00,50.28 C 278.00,50.28 264.28,51.18 264.28,51.18 264.28,51.18 246.00,53.61 246.00,53.61 212.67,59.62 182.18,71.62 154.00,90.52 96.74,128.93 58.68,192.55 51.83,261.00 51.83,261.00 50.00,283.00 50.00,283.00 49.82,307.52 55.46,341.95 63.70,365.00 63.70,365.00 68.40,376.00 68.40,376.00 72.03,384.90 76.17,393.60 80.86,402.00 87.10,413.18 93.92,423.94 98.47,436.00 101.50,444.05 105.63,460.65 113.04,465.07 119.00,468.62 126.30,465.87 130.47,460.91 135.96,454.39 137.26,443.64 142.88,434.00 144.89,430.54 150.24,422.50 155.00,424.17 158.64,425.45 161.31,430.55 162.48,434.00 166.16,444.88 163.96,453.36 166.29,464.00 170.06,481.21 183.72,480.17 193.07,493.00 201.86,505.04 197.24,512.43 200.79,523.00 202.65,528.56 206.80,532.07 212.72,532.52 215.30,532.71 217.75,531.94 219.96,530.62 228.77,525.38 226.14,513.73 227.39,508.00 229.17,499.82 235.89,494.87 244.00,494.19 260.89,492.77 271.19,508.51 274.34,523.00 278.12,540.43 276.86,558.69 275.00,576.28 275.00,576.28 275.00,585.91 275.00,585.91 275.00,585.91 272.00,616.00 272.00,616.00 271.91,624.40 271.41,631.83 276.79,638.96 284.39,649.04 298.62,647.79 304.92,637.00 308.80,630.34 308.01,622.39 308.00,615.00 308.00,615.00 303.00,552.00 303.00,552.00 303.00,552.00 303.00,528.00 303.00,528.00 303.21,508.14 308.40,472.08 334.00,471.04 336.56,470.94 339.52,471.06 342.00,471.71 365.96,477.99 350.95,519.94 377.58,524.43 380.22,524.99 382.38,524.88 385.00,524.43 406.86,519.72 400.29,491.89 404.38,476.00 406.44,468.00 410.96,460.84 420.00,460.11 428.88,459.39 430.50,464.79 430.09,472.09 429.56,481.38 425.44,497.50 437.72,501.23 440.13,501.96 442.71,501.44 444.99,500.51 453.17,497.15 453.06,489.56 453.00,482.00 452.89,470.30 447.46,451.03 457.10,442.22 465.80,434.28 476.70,442.70 483.48,428.00 483.48,428.00 491.92,401.00 491.92,401.00 491.92,401.00 507.66,365.00 507.66,365.00 519.09,332.05 521.14,310.34 521.00,276.00 520.93,259.39 517.49,242.06 513.47,226.00 491.71,138.96 419.06,71.16 331.00,54.74 315.78,51.90 293.33,48.86 278.00,50.28 Z"
          fill="currentColor"
          stroke="inherit"
          strokeWidth="8"
        />
        <rect
          fill="transparent"
          height="100%"
          stroke="currentColor"
          strokeWidth="8"
          width="100%"
        />
      </svg>
    </div>
    <div
      className="absolute top-0 left-0 z-10 h-full w-full text-brown"
    >
      <img
        alt="James Watt"
        className="object-scale-down"
        data-nimg="fill"
        decoding="async"
        onError={[Function]}
        onLoad={[Function]}
        sizes="310px"
        src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
        srcSet="/_next/image?url=%2Fimg.jpg&w=16&q=75 16w, /_next/image?url=%2Fimg.jpg&w=32&q=75 32w, /_next/image?url=%2Fimg.jpg&w=48&q=75 48w, /_next/image?url=%2Fimg.jpg&w=64&q=75 64w, /_next/image?url=%2Fimg.jpg&w=96&q=75 96w, /_next/image?url=%2Fimg.jpg&w=128&q=75 128w, /_next/image?url=%2Fimg.jpg&w=256&q=75 256w, /_next/image?url=%2Fimg.jpg&w=384&q=75 384w, /_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
        style={
          {
            "backgroundImage": "url("data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http%3A//www.w3.org/2000/svg'%3E%3Cimage style='filter:blur(20px)' x='0' y='0' height='100%25' width='100%25' href='data:image/png;base64,imagedata'/%3E%3C/svg%3E")",
            "backgroundPosition": "50% 50%",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "cover",
            "bottom": 0,
            "color": "transparent",
            "height": "100%",
            "left": 0,
            "objectFit": undefined,
            "objectPosition": undefined,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": "100%",
          }
        }
      />
    </div>
    <svg
      aria-label="Water Drop"
      className="absolute left-[216.5px] top-[184px] z-20"
      height="350"
      viewBox="0 0 100 200"
      width="100"
    >
      <circle
        className="fill-white dark:fill-redgrey"
        cx="24"
        cy={10}
        r="6"
      />
    </svg>
    <svg
      aria-label="Growing Drop"
      className="absolute left-[232px] top-[215px] z-20"
      width="16.5"
    >
      <path
        className="fill-white dark:fill-redgrey"
        d="M0,48 a1,1.5 0 0,0 16.5,0"
      />
    </svg>
  </div>
  <svg
    aria-label="Wave Tank"
    className="-mt-5"
    height="100px"
    preserveAspectRatio="none"
    viewBox="0 0 100 100"
    width="100%"
  >
    <polygon
      className="fill-white dark:fill-redgrey dark:stroke-redgrey"
      points="0,100 0,0 0,0 1,0 2,0 3,0 4,0 5,0 6,0 7,0 8,0 9,0 10,0 11,0 12,0 13,0 14,0 15,0 16,0 17,0 18,0 19,0 20,0 21,0 22,0 23,0 24,0 25,0 26,0 27,0 28,0 29,0 30,0 31,0 32,0 33,0 34,0 35,0 36,0 37,0 38,0 39,0 40,0 41,0 42,0 43,0 44,0 45,0 46,0 47,0 48,0 49,0 50,0 51,0 52,0 53,0 54,0 55,0 56,0 57,0 58,0 59,0 60,0 61,0 62,0 63,0 64,0 65,0 66,0 67,0 68,0 69,0 70,0 71,0 72,0 73,0 74,0 75,0 76,0 77,0 78,0 79,0 80,0 81,0 82,0 83,0 84,0 85,0 86,0 87,0 88,0 89,0 90,0 91,0 92,0 93,0 94,0 95,0 96,0 97,0 98,0 99,0 100,0 100,100"
      stroke="white"
      transform="translate(0, 50)"
    />
  </svg>
</section>
`;
